### PR TITLE
Growing coverage for some rules

### PR DESCRIPTION
**Pull request by zxqfox**
I've dropped some checks because I didn't find cases for them.

Strongly need review.

**edit** Locally I see 93% ;-) Here is 95%.

### CODE REVIEW

File: lib/rules/require-curly-braces.js [hunk 1]
```diff
@@ -72,7 +72,7 @@ module.exports.prototype = {
         if (typeIndex['case'] || typeIndex['default']) {
             file.iterateNodesByType('SwitchCase', function(node) {
                 // empty case statement
-                if (!node.consequent || node.consequent.length === 0) {
+                if (node.consequent.length === 0) {
```
Comment (mrjoelkemp) [inline lib/rules/require-curly-braces.js]: In general, my problem with the weakening of the conditions is that they were added for edge cases. If we don't have a test hitting that condition, then we should add the test, not remove the condition. 

With the removal, this would throw an error on empty case statements since `.length` wouldn't exist on `undefined`.

**Comment by mrjoelkemp** [comment]
Actually, those files are already 100% covered according to coveralls. What are you using to see coverage locally?

File: lib/rules/require-curly-braces.js (hunk 1 above)
Comment (zxqfox) [inline lib/rules/require-curly-braces.js]: Parser API said: consequent should exists and be an Array in any cases. So no need to check it.

File: lib/rules/disallow-space-after-keywords.js [hunk 2]
```diff
@@ -31,7 +31,7 @@ module.exports.prototype = {
         file.iterateTokensByType('Keyword', function(token, i, tokens) {
             if (keywordIndex[token.value]) {
                 var nextToken = tokens[i + 1];
-                if (nextToken && nextToken.range[0] !== token.range[1]) {
```
Comment (zxqfox) [inline lib/rules/disallow-space-after-keywords.js]: Here I can't imagine a situation where nextToken for Keyword wouldn't exist. Probably I'm wrong.

File: lib/rules/disallow-space-before-keywords.js [hunk 3]
```diff
@@ -52,7 +52,7 @@ module.exports.prototype = {
                 var diff = token.range[0] - prevToken.range[1];
 
                 if (prevToken.loc.end.line === token.loc.start.line && diff !== 0) {
-                    if (prevToken.type !== 'Punctuator' || prevToken.value !== ';') {
```
Comment (zxqfox) [inline lib/rules/disallow-space-before-keywords.js]: Same as above. Need help with case.

**Comment by zxqfox** [comment]
> Actually, those files are already 100% covered according to coveralls.

I've just ran `npm run coverage-html` ;-)

These conditions marked yellow, and I think this information doesn't placed in `lcov` and these statements marked as covered. But actually it's overkill.

**Comment by mdevils** [comment]
Would you please try and use our assertion methods?

* Assertions: https://github.com/jscs-dev/node-jscs/blob/master/lib/token-assert.js
* Example: https://github.com/jscs-dev/node-jscs/blob/master/lib/rules/disallow-space-before-binary-operators.js

**Comment by zxqfox** [comment]
@mdevils Yeah, why not.

What if there is no `prevToken`? It will be a uncaught error at [lib/token-assert.js#L124](https://github.com/jscs-dev/node-jscs/blob/master/lib/token-assert.js#L124)

**Comment by mdevils** [comment]
Feel free to fix that check.

**Comment by zxqfox** [comment]
@mdevils Okey. I had tried to find test cases before. Can you please provide some ideas how to test these cases.

**Comment by mdevils** [comment]
In this case — no idea.

If you want to test `tokenBefore` with no `prevToken`, you can pass the first document token.

**Comment by zxqfox** [comment]
So I'm not sure but feels like we just don't need that check. ;-\ Okay. I'll rewrite these rules with `token-assert`. Thanks!

**Comment by mdevils** [comment]
That check is needed in `tokenBefore`, not outside of it. So its easy to add and write a test.

File: lib/rules/require-curly-braces.js (hunk 1 above)
Comment (mrjoelkemp) [inline lib/rules/require-curly-braces.js]: Right. My bad, I was thinking about `SwitchStatement`, not `SwitchCase`. This seems okay then.

**Comment by zxqfox** [comment]
@mdevils There is a problem with comments between tokens. Should I inject comments search logic into `.whitespaceBetween` and `.noWhitespaceBetween` functions?

**Comment by mdevils** [comment]
Which problem did you face?

**Comment by zxqfox** [comment]
@mdevils `/* comment */ if /* yolo */ ( /* another */ condition /* more */) /* again */ { // etc`

Should `whitespaceBetween` ignore them? Should it count spaces outside of comments between tokens?

**Comment by zxqfox** [comment]
@mdevils I suppose to use a helpers `getCommentAfter/BeforeToken` to get the closest comment for token. It simplifies a lot inner rule code for a bunch of rules. https://github.com/zxqfox/node-jscs/commit/7a86811c61e29ec8b47cd309a3c6a6bb349441b9

File: lib/js-file.js [hunk 4]
```diff
@@ -151,6 +151,42 @@ JsFile.prototype = {
         this._tokenRangeEndIndex = tokenRangeEndIndex;
     },
     /**
+     * Builds comments index by starting pos for futher navigation.
+     */
+    _buildCommentIndex: function() {
+        var comments = this.getComments();
+        var tokens = this.getTokens();
+        var tokenIterator = 0;
```
Comment (mdevils) [inline lib/js-file.js]: `tokenIterator` is a strange name for this. Let's go for `tokenIndex`.

**Comment by mdevils** [comment]
LGTM.

@markelog, please take a look.

File: lib/js-file.js [hunk 5]
```diff
@@ -368,6 +404,32 @@ JsFile.prototype = {
         });
     },
     /**
+     * Returns comment node directly followed by a token
+     *
+     * @param {Object} token
+     * @returns {Object|undefined}
+     */
+    getCommentAfterToken: function(token) {
+        if (!this._tokenCommentsAfterIndex) {
```
Comment (zxqfox) [inline lib/js-file.js]: Is it fine if we checking it in methods? We can call here `_buildCommentIndex` and move check there. Also we can build that index every time just after token index building.

**Comment by markelog** [comment]
Still a WIP?

File: test/rules/disallow-space-before-keywords.js [hunk 6]
```diff
@@ -45,6 +45,18 @@ describe('rules/dissalow-space-before-keywords', function() {
         assert(checker.checkString('if (true) {\n} /**/else { x++; }').isEmpty());
     });
 
+    it('should not report if tokens placed on different lines', function() {
+        // checking !(prevToken.loc.end.line === token.loc.start.line)
```
Comment (mikesherov) [inline test/rules/disallow-space-before-keywords.js]: Remove this comment.

File: test/rules/disallow-space-before-keywords.js [hunk 7]
```diff
@@ -45,6 +45,18 @@ describe('rules/dissalow-space-before-keywords', function() {
...
+        checker.configure({ disallowSpaceBeforeKeywords: true });
+        assert(checker.checkString('x\nif (true) false').isEmpty());
+    });
+
+    it('should not report if there are no braces', function() {
+        // checking !(prevToken.type !== 'Punctuator' || prevToken.value !== ';')
```
Comment (mikesherov) [inline test/rules/disallow-space-before-keywords.js]: Remove comment.

File: test/rules/require-space-before-keywords.js [hunk 8]
```diff
@@ -64,6 +64,18 @@ describe('rules/require-space-before-keywords', function() {
         assert(errors.explainError(error).indexOf('Should be one space instead of 2, before "else"') === 0);
     });
 
+    it('should not trigger error for different lines', function() {
+        // !(prevToken.loc.end.line === token.loc.start.line)
```
Comment (mikesherov) [inline test/rules/require-space-before-keywords.js]: Remove

File: test/rules/require-space-before-keywords.js [hunk 9]
```diff
@@ -64,6 +64,18 @@ describe('rules/require-space-before-keywords', function() {
...
+        checker.configure({ requireSpaceBeforeKeywords: ['else'] });
+        assert(checker.checkString('if (true) {\n} /**/ \nelse { x++; }').isEmpty());
+    });
+
+    it('should not report if there are no braces', function() {
+        // checking !(prevToken.type !== 'Punctuator' || prevToken.value !== ';')
```
Comment (mikesherov) [inline test/rules/require-space-before-keywords.js]: Remove.

**Comment by zxqfox** [comment]
Fixed last nuts. Thanks for review, guys!

/cc @markelog

**Comment by markelog** [comment]
Cudos for this hard work!
